This is work of Alina Dzemiantsevich
ademencevich@gmail.com

## Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

### Step 1. Open the data file and have a look at the general information. 

In [1]:
import pandas as pd
#data = pd.read_csv('/datasets/credit_scoring_eng.csv')
data = pd.read_csv('credit_scoring_eng.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [2]:
data.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate

### Conclusion

We have a dataframe which contain information about potential customers of the bank. Looking through the general information of the dataframe we see that not all columns have the same number of elements. The max number of elements is 21525. The vast majority of columns has such amount of elements. But there are _'days_employed'_ and _'total_income'_ columns with smaller number of elements. It says these columns most likely contain missing values. To identify which exactly mistakes and problems with the database we have, we should check data for Nan and None Values and look for unique elements in columns. 
In column _'days_employed'_ we also can see lots of entries with negative data as well as we have there Nan values, so we have to investigate the reason of it and what this data is showing us.

### Step 2. Data preprocessing

### Processing missing values

Let's begin analyzing data with _'days_employed'_ column as this is  one of two columns were number of entries smaller than in other columns.

We have 2174 rows with NaN Data in _'days_employed'_ and _'total_income'_ columns

In [3]:
data[data['days_employed'].isna()].count()

children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

In [4]:
data[data['days_employed'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

Among entries with NaN value in _'total_income'_ and _'days_employed'_ columns there are 170 people, who defaulted on a loan

In [5]:
data[data['days_employed'].isna()]['debt'].value_counts()

0    2004
1     170
Name: debt, dtype: int64

I don't think that we should just delete rows which contain NaN values in _'days_employed'_ and _'total_income'_. These rows contain valuable information in all the others columns. 2174 rows is 1% of all database. It is not a huge part of database. So we can leave it for now.

In [6]:
data['days_employed'].mean()

63046.49766147338

In [7]:
median_days_employed = data['days_employed'].median()
print(median_days_employed)

-1203.369528770489


I will use median value for filling the missing values because it is the value which divide dataset to higher half from the lower half of the data.
It is more close to the real numbers than ordinary average.

In [8]:
#Fill NaN values of 'days_employed' with median value for the 'days_employed' column
data['days_employed'] = data['days_employed'].fillna(value = median_days_employed)

Now we see here that all entries with NaN value in 'days_employed' column were replaced by median of the column. Let's do the same for _'total_income'_ column

In [9]:
data[data['total_income'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0   -1203.369529         65  secondary education             1   
26            0   -1203.369529         41  secondary education             1   
29            0   -1203.369529         63  secondary education             1   
41            0   -1203.369529         50  secondary education             1   
55            0   -1203.369529         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2   -1203.369529         47  Secondary Education             1   
21495         1   -1203.369529         50  secondary education             1   
21497         0   -1203.369529         48    BACHELOR'S DEGREE             0   
21502         1   -1203.369529         42  secondary education             1   
21510         2   -1203.369529         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [10]:
data['total_income'].mean()

26787.56835465867

In [11]:
median_total_income = data['total_income'].median()
print(median_total_income)

23202.87


Here in the _'total_income'_ column values of median and mean are quite close. But I also will use median value as more precise to fill the NaNs.

In [12]:
#Fill NaN values of 'total income' column with median value of the column.
data['total_income'] = data['total_income'].fillna(value = median_total_income)

So after that we don't have missing values in the column.

In [13]:
data[data['total_income'].isna()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

What we have:

In [14]:
data[data['total_income']== median_total_income]

children  days_employed  dob_years            education  education_id  \
12            0   -1203.369529         65  secondary education             1   
26            0   -1203.369529         41  secondary education             1   
29            0   -1203.369529         63  secondary education             1   
41            0   -1203.369529         50  secondary education             1   
55            0   -1203.369529         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2   -1203.369529         47  Secondary Education             1   
21495         1   -1203.369529         50  secondary education             1   
21497         0   -1203.369529         48    BACHELOR'S DEGREE             0   
21502         1   -1203.369529         42  secondary education             1   
21510         2   -1203.369529         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12         23202.87         to have a wedding  
26         23202.87                 education  
29         23202.87    building a real estate  
41         23202.87  second-hand car purchase  
55         23202.87         to have a wedding  
...             ...                       ...  
21489      23202.87         purchase of a car  
21495      23202.87          wedding ceremony  
21497      23202.87       building a property  
21502      23202.87    building a real estate  
21510      23202.87              car purchase  

[2175 rows x 12 columns]

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Now we see that we don't have so obvious missing values. Each column contain 21525 non-null values. 

So let's take a look on column with customers age which is going after _'days_employed'_ column. 

In [16]:
print(data['dob_years'].unique())
print(data['dob_years'].min())

[42 36 33 32 53 27 43 50 35 41 40 65 54 56 26 48 24 21 57 67 28 63 62 47
 34 68 25 31 30 20 49 37 45 61 64 44 52 46 23 38 39 51  0 59 29 60 55 58
 71 22 73 66 69 19 72 70 74 75]
0


In [17]:
data.loc[data['dob_years'] == 0].count()

children            101
days_employed       101
dob_years           101
education           101
education_id        101
family_status       101
family_status_id    101
gender              101
income_type         101
debt                101
total_income        101
purpose             101
dtype: int64

In [18]:
#We see there are clients with "0" age
data.loc[data['dob_years'] == 0]

children  days_employed  dob_years            education  education_id  \
99            0  346541.618895          0  Secondary Education             1   
149           0   -2664.273168          0  secondary education             1   
270           3   -1872.663186          0  secondary education             1   
578           0  397856.565013          0  secondary education             1   
1040          0   -1158.029561          0    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
19829         0   -1203.369529          0  secondary education             1   
20462         0  338734.868540          0  secondary education             1   
20577         0  331741.271455          0  secondary education             1   
21179         2    -108.967042          0    bachelor's degree             0   
21313         0   -1268.487728          0  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
99          married                 0      F     retiree     0     11406.644   
149        divorced                 3      F    employee     0     11228.230   
270         married                 0      F    employee     0     16346.633   
578         married                 0      F     retiree     0     15619.310   
1040       divorced                 3      F    business     0     48639.062   
...             ...               ...    ...         ...   ...           ...   
19829       married                 0      F    employee     0     23202.870   
20462       married                 0      F     retiree     0     41471.027   
20577     unmarried                 4      F     retiree     0     20766.202   
21179       married                 0      M    business     0     38512.321   
21313     unmarried                 4      M    employee     0     24387.070   

                            purpose  
99                              car  
149            housing transactions  
270              housing renovation  
578    construction of own property  
1040                   to own a car  
...                             ...  
19829                       housing  
20462      purchase of my own house  
20577                      property  
21179        building a real estate  
21313          housing transactions  

[101 rows x 12 columns]

As we can see here there are 101 customers of 0 age. This is strange, especially because we have non null values of _'days_employed'_ in these the same rows  (but we still have to process this column _'days_employed'_ to understand what these numbers mean). I will leave _'dob_years'_ with null values for now, because we will not use this information about age of the customers in our conclusions, no one question is touching this data about client age. But we will keep in mind that we have such a strange data here.

### Conclusion

As we can see now we don't have any obvious missing values. After the export of the database there were 2174 missing values both in _'days_employed'_ and _'total_income'_ columns. NaN values were replaced by median value for each column.

### Data type replacement

Let's find how many years each client of the bank worked

In [19]:
#First of all convert real data into integer
data['days_employed'] = data['days_employed'].astype('int')

In [20]:
#Let's define constant with number of days in the year 
#and count how many years each client has worked

days_in_a_year = 365
data['years_employed'] = data['days_employed'].agg(lambda x: abs(x) / days_in_a_year)

In [21]:
#Check what we got
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  Secondary Education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed  
0    purchase of the house       23.115068  
1             car purchase       11.024658  
2    purchase of the house       15.405479  
3  supplementary education       11.298630  
4        to have a wedding      932.235616

In [22]:
#Change type of new column to int for more clear view
data['years_employed'] = data['years_employed'].astype('int')

In [23]:
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  Secondary Education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed  
0    purchase of the house              23  
1             car purchase              11  
2    purchase of the house              15  
3  supplementary education              11  
4        to have a wedding             932

In [24]:
#Check what we got: days_employed     21525 non-null  int64 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  int64  
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
 12  years_employed    21525 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 2.1+ MB


### Conclusion

I replaced the real number data type with the integer type in _'days_employed'_ column. There were negative numbers which I changed using abs function and dividing by numbers of days in the year. I got the data of how many years each client of the bank has worked. 

I can make an conclusion that the real data can be tricky and messy. But behind any number even with error stands information which can be useful.

### Processing duplicates

Let's see which and how many unique values we have in other dataframe columns.

In [25]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

Here we can see strange values of having -1 and 20 kids. Let's see how many such values do we have.

In [26]:
data.loc[data['children'] == 20].count()

children            76
days_employed       76
dob_years           76
education           76
education_id        76
family_status       76
family_status_id    76
gender              76
income_type         76
debt                76
total_income        76
purpose             76
years_employed      76
dtype: int64

In [27]:
data.loc[data['children'] == -1].count()

children            47
days_employed       47
dob_years           47
education           47
education_id        47
family_status       47
family_status_id    47
gender              47
income_type         47
debt                47
total_income        47
purpose             47
years_employed      47
dtype: int64

Looking on values of _'children'_ column we see that we don't have NaN values, but there are unusual values of -1 and 20 children.

I understand that we are living in Israel and things like 20 children in a family might happen :) But I think the most likely there were made a mistake and might be that entry of 20 children was supposed to be 2 children. Only 76 rows contain this value of 20 children among 21525 entries. This is approximately 0,35% of data. I think we can just delete these rows without loosing of valuable information. 

According to the data we have 47 rows with entry of -1 child. I think this also might be mistake and has to be 1 children or maybe this means woman pregnancy? We can't check this hypothesis (about pregnancy). So I would also delete these rows.

In [28]:
#Let's see before deleting how many people with these number of children defaulted a loan.
data.loc[(data['children'] == 20) & (data['debt'] ==1)]

children  days_employed  dob_years            education  education_id  \
1074         20          -3310         56  secondary education             1   
10782        20          -1044         34  Secondary Education             1   
12319        20           -507         50  Secondary Education             1   
12534        20           -579         48  secondary education             1   
18890        20           -904         37  secondary education             1   
20038        20           -782         32  secondary education             1   
20355        20          -2695         45  secondary education             1   
21008        20          -1240         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
1074             married                 0      F    employee     1   
10782           divorced                 3      M    employee     1   
12319  civil partnership                 1      M    business     1   
12534  civil partnership                 1      F    employee     1   
18890            married                 0      F    business     1   
20038          unmarried                 4      M    employee     1   
20355  civil partnership                 1      M    employee     1   
21008            married                 0      F    employee     1   

       total_income                       purpose  years_employed  
1074      36722.966          getting an education               9  
10782     23127.125  construction of own property               2  
12319     27164.944              having a wedding               1  
12534     20082.890              having a wedding               1  
18890     34049.851      buying a second-hand car               2  
20038     18841.368    buy commercial real estate               2  
20355     28467.621              having a wedding               7  
21008     21363.842                  to own a car               3

In [29]:
#Looks like 8 persons with 20 children and 1 person with -1 child defaulted a loan
data.loc[(data['children'] == -1) & (data['debt'] ==1)]

children  days_employed  dob_years            education  education_id  \
16129        -1           -457         33  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
16129       married                 0      F    employee     1     23942.573   

      purpose  years_employed  
16129     car               1

In [30]:
#Delete rows with 20 children 
data.drop(data[data['children'] == 20].index, inplace=True)

In [31]:
#76 rows with data of 20 children was deleted
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21449 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21449 non-null  int64  
 1   days_employed     21449 non-null  int64  
 2   dob_years         21449 non-null  int64  
 3   education         21449 non-null  object 
 4   education_id      21449 non-null  int64  
 5   family_status     21449 non-null  object 
 6   family_status_id  21449 non-null  int64  
 7   gender            21449 non-null  object 
 8   income_type       21449 non-null  object 
 9   debt              21449 non-null  int64  
 10  total_income      21449 non-null  float64
 11  purpose           21449 non-null  object 
 12  years_employed    21449 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 2.3+ MB


In [32]:
#Delete rows with -1 children 
data.drop(data[data['children'] == -1].index, inplace=True)

In [33]:
#47 rows with data of -1 child was deleted
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21402 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21402 non-null  int64  
 1   days_employed     21402 non-null  int64  
 2   dob_years         21402 non-null  int64  
 3   education         21402 non-null  object 
 4   education_id      21402 non-null  int64  
 5   family_status     21402 non-null  object 
 6   family_status_id  21402 non-null  int64  
 7   gender            21402 non-null  object 
 8   income_type       21402 non-null  object 
 9   debt              21402 non-null  int64  
 10  total_income      21402 non-null  float64
 11  purpose           21402 non-null  object 
 12  years_employed    21402 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 2.3+ MB


In [34]:
#now it looks much better
data['children'].value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

In [35]:
len(data['children'])

21402

Let's look now on _'education'_ column

In [36]:
data['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

Data in this column is quite messy. We see "bachelor's degree" and "Bachelor's Degree" which is the same. So let's make strings in a column '_education'_ lowercase.

In [37]:
#Making strings in a column 'education' lowercase
data['education'] = data['education'].str.lower()
#check for unique values
data['education'].unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

Now it looks more clean.
Let's check _'family_status'_ column

In [38]:
data['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

In column _'family_status'_ everything looks clean and correct

In [39]:
data['family_status'].value_counts()

married              12302
civil partnership     4160
unmarried             2799
divorced              1189
widow / widower        952
Name: family_status, dtype: int64

In [40]:
data['family_status_id'].unique()

array([0, 1, 2, 3, 4])

In [41]:
#the numbers of 'family_status_id' correspond with the numbers of 'family_status' column
data['family_status_id'].value_counts()

0    12302
1     4160
4     2799
3     1189
2      952
Name: family_status_id, dtype: int64

Let's look on values of _'gender'_ column. We can see one entry with XNA gender. Assuming that this person can be transgender person leave it as it is.

In [42]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [43]:
data['gender'].value_counts()

F      14154
M       7247
XNA        1
Name: gender, dtype: int64

In [44]:
data[data['gender'] == 'XNA']

children  days_employed  dob_years     education  education_id  \
10701         0          -2358         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  years_employed  
10701     32624.825  buy real estate               6

In [45]:
#Cheack for unique values of 'income_type' column
data['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

In [46]:
data['income_type'].value_counts()

employee                       11050
business                        5054
retiree                         3839
civil servant                   1453
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

I don't see the big difference between 'business' and 'entrepreneur' category in _'income_type'_ column. Let's combine these groups. There is point especially because 'entrepreneur' category contain only two entries.

In [47]:
data[data['income_type'] == 'entrepreneur']

children  days_employed  dob_years          education  education_id  \
5936          0          -1203         58  bachelor's degree             0   
18697         0           -520         27  bachelor's degree             0   

           family_status  family_status_id gender   income_type  debt  \
5936             married                 0      M  entrepreneur     0   
18697  civil partnership                 1      F  entrepreneur     0   

       total_income                      purpose  years_employed  
5936      23202.870  buy residential real estate               3  
18697     79866.103             having a wedding               1

In [48]:
data.loc[(data['income_type'] == 'entrepreneur'),'income_type'] = 'business'

In [49]:
data['income_type'].value_counts()

employee                       11050
business                        5056
retiree                         3839
civil servant                   1453
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

**Let's check finally the dataframe for duplicate rows.**

In [50]:
data.duplicated().sum()

71

In [51]:
data[data.duplicated()].head(5)

children  days_employed  dob_years            education  education_id  \
2849         0          -1203         41  secondary education             1   
3290         0          -1203         58  secondary education             1   
4182         1          -1203         34    bachelor's degree             0   
4851         0          -1203         60  secondary education             1   
5557         0          -1203         58  secondary education             1   

          family_status  family_status_id gender income_type  debt  \
2849            married                 0      F    employee     0   
3290  civil partnership                 1      F     retiree     0   
4182  civil partnership                 1      F    employee     0   
4851  civil partnership                 1      F     retiree     0   
5557  civil partnership                 1      F     retiree     0   

      total_income                              purpose  years_employed  
2849      23202.87  purchase of the house for my family               3  
3290      23202.87                    to have a wedding               3  
4182      23202.87                     wedding ceremony               3  
4851      23202.87                     wedding ceremony               3  
5557      23202.87                    to have a wedding               3

We have 71 rows with duplicate data. So let's delete it from the dataset. 

In [52]:
data.drop_duplicates(inplace = True)

In [53]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21331 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21331 non-null  int64  
 1   days_employed     21331 non-null  int64  
 2   dob_years         21331 non-null  int64  
 3   education         21331 non-null  object 
 4   education_id      21331 non-null  int64  
 5   family_status     21331 non-null  object 
 6   family_status_id  21331 non-null  int64  
 7   gender            21331 non-null  object 
 8   income_type       21331 non-null  object 
 9   debt              21331 non-null  int64  
 10  total_income      21331 non-null  float64
 11  purpose           21331 non-null  object 
 12  years_employed    21331 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 2.3+ MB


We had dataframe with 21402 rows and now after deleting 71 rows of duplicate data we have 21331 rows.

### Conclusion

I had checked dataset for duplicates and odd values. Where it was possible I deleted rows with such values (like number of 20 and -1 children), in some places I replaced the values, in some left as it is. I also checked the dataset for duplicates (71 rows) and deleted them.

### Categorizing Data

Here in the 'purpose' column are presented the purposes of the clients loan. This data  look quite messy. To answer the set question **How do different loan purposes affect on-time loan repayment?**, we should categorize the data and to lower the number of purposes.

In [54]:
data['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

With the help of value_counts we can see some patterns in purposes of loan but still we should process these groups.

In [55]:
data['purpose'].value_counts()

wedding ceremony                            790
having a wedding                            763
to have a wedding                           760
real estate transactions                    672
buy commercial real estate                  658
buying property for renting out             649
housing transactions                        647
transactions with commercial real estate    645
housing                                     641
purchase of the house                       640
purchase of the house for my family         637
property                                    631
construction of own property                628
transactions with my real estate            623
building a real estate                      620
building a property                         619
purchase of my own house                    619
buy real estate                             616
housing renovation                          604
buy residential real estate                 602
buying my own car                       

In [56]:
#Let's try to find the main purposes of the loan among given 
import nltk
nltk.download('punkt')
from nltk.stem import WordNetLemmatizer
wordnet_lemma = WordNetLemmatizer() 

[nltk_data] Downloading package punkt to /Users/lin/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [57]:
from nltk.tokenize import word_tokenize
data['purpose_tokenized'] = data['purpose'].apply(word_tokenize)

In [58]:
data['purpose_tokenized'].head(5)

0    [purchase, of, the, house]
1               [car, purchase]
2    [purchase, of, the, house]
3    [supplementary, education]
4        [to, have, a, wedding]
Name: purpose_tokenized, dtype: object

In [59]:
data['purpose_tokenized'].value_counts()

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 1709, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[wedding, ceremony]                               790
[having, a, wedding]                              763
[to, have, a, wedding]                            760
[real, estate, transactions]                      672
[buy, commercial, real, estate]                   658
[buying, property, for, renting, out]             649
[housing, transactions]                           647
[transactions, with, commercial, real, estate]    645
[housing]                                         641
[purchase, of, the, house]                        640
[purchase, of, the, house, for, my, family]       637
[property]                                        631
[construction, of, own, property]                 628
[transactions, with, my, real, estate]            623
[building, a, real, estate]                       620
[purchase, of, my, own, house]                    619
[building, a, property]                           619
[buy, real, estate]                               616
[housing, renovation]       

In [60]:
"""Actually I wanted to count the words from 'purpose_tokenized' column or just 'purpose'
which appears the most in dataframe column
but I didn't found how to do this for the column not just for rows
so because of it I leave it commented

from collections import Counter
data['purp_count'] = data['purpose_tokenized'].apply(Counter)
print(data['purp_count'])"""

"Actually I wanted to count the words from 'purpose_tokenized' column or just 'purpose'\nwhich appears the most in dataframe column\nbut I didn't found how to do this for the column not just for rows\nso because of it I leave it commented\n\nfrom collections import Counter\ndata['purp_count'] = data['purpose_tokenized'].apply(Counter)\nprint(data['purp_count'])"

In [61]:
#let's write a function to categorize the data
def purpose_loan(string):
    if 'wedding' in string:
        return 'wedding'
    if 'real estate' in string:
        return 'property'
    if 'property' in string:
        return 'property'
    if 'hous' in string:
        return 'property'
    if 'car' in string:
        return 'car'
    if 'university' in string:
        return 'education'
    if 'educat' in string:
        return 'education'

In [62]:
#apply the function and categorize the data by purposes into 4 categories
data['purpose_groupped'] = data['purpose'].apply(purpose_loan)

In [63]:
#cheack what we got
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  secondary education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed           purpose_tokenized  \
0    purchase of the house              23  [purchase, of, the, house]   
1             car purchase              11             [car, purchase]   
2    purchase of the house              15  [purchase, of, the, house]   
3  supplementary education              11  [supplementary, education]   
4        to have a wedding             932      [to, have, a, wedding]   

  purpose_groupped  
0         property  
1              car  
2         property  
3        education  
4          wedding

In [64]:
#Now we have 4 categories for the loan purposes
data['purpose_groupped'].value_counts()

property     10751
car           4279
education     3988
wedding       2313
Name: purpose_groupped, dtype: int64

In [65]:
#also categorize the data according the data of whether the customer
#has ever defaulted a loan
def isdefolted(x):
    if x ==0:
        return 'never defaulted a loan'
    if x ==1:
        return 'defaulted a loan before'

In [66]:
#apply the function
data['defaulted_a_loan'] = data['debt'].apply(isdefolted)

In [67]:
#check what we got
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  secondary education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed           purpose_tokenized  \
0    purchase of the house              23  [purchase, of, the, house]   
1             car purchase              11             [car, purchase]   
2    purchase of the house              15  [purchase, of, the, house]   
3  supplementary education              11  [supplementary, education]   
4        to have a wedding             932      [to, have, a, wedding]   

  purpose_groupped        defaulted_a_loan  
0         property  never defaulted a loan  
1              car  never defaulted a loan  
2         property  never defaulted a loan  
3        education  never defaulted a loan  
4          wedding  never defaulted a loan

In [68]:
#just curious the overall numbers
data['defaulted_a_loan'].value_counts()

never defaulted a loan     19599
defaulted a loan before     1732
Name: defaulted_a_loan, dtype: int64

To answer the second question <b>"Is there a relation between income level and repaying a loan on time?"</b> let's dive into dataset to find which income levels we have there.

First of all find min, max and median value of <i>'total_income'</i>.

In [69]:
print('Max total income',data['total_income'].max())
print('Min total income',data['total_income'].min())
print('Median total income',data['total_income'].median())

Max total income 362496.645
Min total income 3306.762
Median total income 23202.87


We can build three groups of income level now:
- income less than median
- median income
- income grater than median

In [70]:
median_total_income = data['total_income'].median()
print('Median value for the column total income is ', median_total_income)

num1 = len(data[(data['total_income'] < median_total_income)])
num2 = len(data[(data['total_income'] == median_total_income)])
num3 = len(data[(data['total_income'] > median_total_income)])
print('Number of clients with income less than median',num1)
print('Number of clients with income equal to the median',num2)
print('Number of clients with income greater than median', num3)

Median value for the column total income is  23202.87
Number of clients with income less than median 9619
Number of clients with income equal to the median 2092
Number of clients with income greater than median 9620


We have 9619 entries or clients with <i>'total_income'</i> less than meadian value.

And we have 2092 clients with <i>'total_income'</i> equal to median value. We have such a number because these rows were NaN values and we replaced missing values with median and deleted some of duplicates.

Finally we have 9620 clients with <i>'total_income'</i> greater than median value. 

In [71]:
#function to identify each client to one of income level groups
def income_level(income):
    median_total_income = data['total_income'].median()
    if income <  median_total_income:
        return 'smaller income'
    if income == median_total_income:
        return 'median income'
    if income > median_total_income:
        return 'greater income'

In [72]:
#apply the function to the dataset
data['income_level'] = data['total_income'].apply(income_level)

In [73]:
#let's check what we got
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  secondary education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed           purpose_tokenized  \
0    purchase of the house              23  [purchase, of, the, house]   
1             car purchase              11             [car, purchase]   
2    purchase of the house              15  [purchase, of, the, house]   
3  supplementary education              11  [supplementary, education]   
4        to have a wedding             932      [to, have, a, wedding]   

  purpose_groupped        defaulted_a_loan    income_level  
0         property  never defaulted a loan  greater income  
1              car  never defaulted a loan  smaller income  
2         property  never defaulted a loan  greater income  
3        education  never defaulted a loan  greater income  
4          wedding  never defaulted a loan  greater income

In [74]:
#everything looks correct
data['income_level'].value_counts()

greater income    9620
smaller income    9619
median income     2092
Name: income_level, dtype: int64

To categorize data more detailed, we can build 5 categories by income level instead of 3. 
For that we can use quartiles:

In [75]:
#median for total income column
median_total_income = data['total_income'].median()
print('Median for the column total income', median_total_income)

#first quartile
first_quartile = data[data['total_income'] < (data['total_income'].median())]['total_income'].median()
print('First quartile',first_quartile)

#third quartile
third_quartile = data[data['total_income'] > (data['total_income'].median())]['total_income'].median()
print('Third quartile', third_quartile)

#count the number of clients of first quartile
qua1 = len(data[data['total_income'] <= first_quartile])
print('Number of clients of first quartile', qua1)

#count the number of clients of second quartile
qua2 = len(
data[(data['total_income'] > first_quartile) & (data['total_income'] <median_total_income)]
)
print('Number of clients with income of second quartile',qua2)

#count the number of clients with income equal to median value 
qua3 = len(data[data['total_income'] == data['total_income'].median()])
print('Number of clients with income equal to median value of income',qua3)

#count the number of clients of third quartile
qua4 = len(
data[(data['total_income'] > median_total_income) & (data['total_income'] <third_quartile)]
)
print('Number of clients with income of third quartile',qua4)

#count the number of clients of forth quartile
qua5 = len(data[data['total_income'] >= third_quartile])
print('Number of clients with income of forth quartile',qua5)


Median for the column total income 23202.87
First quartile 16479.958
Third quartile 32551.225
Number of clients of first quartile 4810
Number of clients with income of second quartile 4809
Number of clients with income equal to median value of income 2092
Number of clients with income of third quartile 4810
Number of clients with income of forth quartile 4810


In [76]:
#let's write a function which categorizes income levels by quartiles
def income_level_quartiles(income):
    median_total_income = data['total_income'].median()
    first_quartile = data[data['total_income'] < (data['total_income'].median())]['total_income'].median()
    third_quartile = data[data['total_income'] > (data['total_income'].median())]['total_income'].median()
  
    #smallest income - 1st quartile
    #less than median - 2nd quartile
    #median income
    #greater than median - 3rd quartile
    #greatest income - 4th quartile

    if income <=  first_quartile:
        return '1st quartile'
    if (income > first_quartile and income < median_total_income):
        return '2nd quartile'
    if income == median_total_income:
        return 'median income'
    if (income > median_total_income and income < third_quartile):
        return '3rd quartile'
    if income >= median_total_income:
        return '4th quartile'
    

In [77]:
#apply function to the data
data['income_level_quartiles'] = data['total_income'].apply(income_level_quartiles)

In [78]:
#let's check what we got
data.head()

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  secondary education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed           purpose_tokenized  \
0    purchase of the house              23  [purchase, of, the, house]   
1             car purchase              11             [car, purchase]   
2    purchase of the house              15  [purchase, of, the, house]   
3  supplementary education              11  [supplementary, education]   
4        to have a wedding             932      [to, have, a, wedding]   

  purpose_groupped        defaulted_a_loan    income_level  \
0         property  never defaulted a loan  greater income   
1              car  never defaulted a loan  smaller income   
2         property  never defaulted a loan  greater income   
3        education  never defaulted a loan  greater income   
4          wedding  never defaulted a loan  greater income   

  income_level_quartiles  
0           4th quartile  
1           2nd quartile  
2           3rd quartile  
3           4th quartile  
4           3rd quartile

In [79]:
# to check dependencies between marital status and loan payments 
# I want to aggregate some of family_status categories into two by the logic:
# 'single' category includes unmarried people, divorced and widow/ widowers
# 'married' category includes married people and civil partnership

def marital_status(status):
    if status == 'married':
        return 'married/civil partnership'
    if 'civil' in status:
        return 'married/civil partnership'
    if status == 'unmarried':
        return 'single'
    if 'divorced' in status:
        return 'single'
    if 'widow/ widow' in status:
        return 'single'

In [80]:
#apply the categorization function
data['marital_status'] = data['family_status'].apply(marital_status)

In [81]:
#let's check what we got
data.head()

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  secondary education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed           purpose_tokenized  \
0    purchase of the house              23  [purchase, of, the, house]   
1             car purchase              11             [car, purchase]   
2    purchase of the house              15  [purchase, of, the, house]   
3  supplementary education              11  [supplementary, education]   
4        to have a wedding             932      [to, have, a, wedding]   

  purpose_groupped        defaulted_a_loan    income_level  \
0         property  never defaulted a loan  greater income   
1              car  never defaulted a loan  smaller income   
2         property  never defaulted a loan  greater income   
3        education  never defaulted a loan  greater income   
4          wedding  never defaulted a loan  greater income   

  income_level_quartiles             marital_status  
0           4th quartile  married/civil partnership  
1           2nd quartile  married/civil partnership  
2           3rd quartile  married/civil partnership  
3           4th quartile  married/civil partnership  
4           3rd quartile  married/civil partnership

### Conclusion

I categorized data by income level, by purpose of the loan, by that fact if the client has ever defaulted a loan and by marital status. Now we have columns where each client has label according to category which he is belong to. We can use these data to analyze different groups of clients.

### Step 3. Answer these questions

- **Is there a relation between having kids and repaying a loan on time?**

Before we start to answer the question, let's count overall average default rate for the dataset.

In [82]:
average_default_rate = (data['debt'].sum() / len(data['debt']))*100
print('Average default rate for the whole data set is {:.2f}%'.format(average_default_rate))

Average default rate for the whole data set is 8.12%


Now let's answer the set question.

In [83]:
#Let's build a pivot table to see the number of people defaulted a loan before and
#never defaulted a loan

dp_having_a_kids = data.pivot_table(index=['children'],columns='defaulted_a_loan', values='debt', aggfunc='count')
print(dp_having_a_kids)

defaulted_a_loan  defaulted a loan before  never defaulted a loan
children                                                         
0                                  1063.0                 13028.0
1                                   444.0                  4364.0
2                                   194.0                  1858.0
3                                    27.0                   303.0
4                                     4.0                    37.0
5                                     NaN                     9.0


These absolute numbers don't tell us if there are relation between having kids and repaying a loan on time. So let's count percentage of people defaulted a loan and never defaulted for each group - with 0 kids and with 1 kid etc.

In [84]:
#percentage_of_defaulted
dp_having_a_kids['percentage_of_defaulted'] = (dp_having_a_kids['defaulted a loan before']/(dp_having_a_kids['defaulted a loan before'] + dp_having_a_kids['never defaulted a loan']))*100
#percentage of never defaulted or paying
dp_having_a_kids['percentage_of_paying'] = 100 - dp_having_a_kids['percentage_of_defaulted']

#ratio of number of people never defaulted a loan before 
#to thenumber of people who defaulted
dp_having_a_kids['ratio paying to defaulted'] = dp_having_a_kids['never defaulted a loan'] / dp_having_a_kids['defaulted a loan before']

print(dp_having_a_kids)

defaulted_a_loan  defaulted a loan before  never defaulted a loan  \
children                                                            
0                                  1063.0                 13028.0   
1                                   444.0                  4364.0   
2                                   194.0                  1858.0   
3                                    27.0                   303.0   
4                                     4.0                    37.0   
5                                     NaN                     9.0   

defaulted_a_loan  percentage_of_defaulted  percentage_of_paying  \
children                                                          
0                                7.543822             92.456178   
1                                9.234609             90.765391   
2                                9.454191             90.545809   
3                                8.181818             91.818182   
4                                9.756098    

In [85]:
print('Average percentage of paying off the loan by this classification {:.2f}%'.format(dp_having_a_kids['percentage_of_paying'].mean())) 
print('Average percentage of paying off the loan with 1-5 kids {:.2f}%'.format(dp_having_a_kids['percentage_of_paying'][1:].mean()))
print('Percentage of paying off the loan with no kids {:.2f}%'.format(dp_having_a_kids['percentage_of_paying'][0]))
print()
print('Average percentage of defaulted the loan by this classification {:.2f}%'.format(dp_having_a_kids['percentage_of_defaulted'].mean()))
print('Average percentage of defaulted the loan with 1-5 kids {:.2f}%'.format(dp_having_a_kids['percentage_of_defaulted'][1:].mean()))
print('Percentage of defaulted the loan with no kids {:.2f}%'.format(dp_having_a_kids['percentage_of_defaulted'][0]))
print('Max percentage of defaulted the loan {:.2f}%'.format(dp_having_a_kids['percentage_of_defaulted'].max()))


Average percentage of paying off the loan by this classification 91.17%
Average percentage of paying off the loan with 1-5 kids 90.84%
Percentage of paying off the loan with no kids 92.46%

Average percentage of defaulted the loan by this classification 8.83%
Average percentage of defaulted the loan with 1-5 kids 9.16%
Percentage of defaulted the loan with no kids 7.54%
Max percentage of defaulted the loan 9.76%


### Conclusion

In this pivot table and inputs above we see that percentage of people who have 0 children and always pay off a loan is **92.46%** (default rate is 7.54%). 

At the same time the mean percentage of people who have from 1 to 5 children and always pay off a loan is **90,84%** (default rate is 9.16%). 

We also can see that for each category percentage of people who pay off a loan is approximately from **90% to 92%**.

So the difference in numbers is not so big. Average percentage of paying off the loan is same for all groups of people independently of number of children.

The smallest percentage of people who defaulted a loan is in group where are no kids and where are 3 children in a family. These people will default a loan in **7,54%** and **8,18%** cases accordingly. In other cases - **having 1, 2 or 4 kids, the percentage to default a loan is approximately 9%**. 

- Maximum risk is to give a loan to the client with 4 kids - 9,75% of them not return.
- People with 5 kids according to our data have never defaulted a loan.

- People with no kids are the most trustable for us.
- People with 1, 2 and 4 kids - less.

- **Is there a relation between marital status and repaying a loan on time?**

In [86]:
#To answer the question let's build a pivot table, using the column 'family_status'
#name dp_ms is for data pivot marital status
dp_ms = data.pivot_table(index=['family_status'],columns='defaulted_a_loan', values='debt', aggfunc='count')

dp_ms['percentage_of_defaulted']=(dp_ms['defaulted a loan before']/(dp_ms['never defaulted a loan']+dp_ms['defaulted a loan before']))*100
dp_ms['percentage_of_paying'] = 100 - dp_ms['percentage_of_defaulted']
dp_ms['ratio paying to defaulted'] = dp_ms['never defaulted a loan']/dp_ms['defaulted a loan before']

print(dp_ms)

defaulted_a_loan   defaulted a loan before  never defaulted a loan  \
family_status                                                        
civil partnership                      385                    3749   
divorced                                84                    1105   
married                                927                   11334   
unmarried                              273                    2523   
widow / widower                         63                     888   

defaulted_a_loan   percentage_of_defaulted  percentage_of_paying  \
family_status                                                      
civil partnership                 9.313014             90.686986   
divorced                          7.064760             92.935240   
married                           7.560558             92.439442   
unmarried                         9.763948             90.236052   
widow / widower                   6.624606             93.375394   

defaulted_a_loan   ratio paying 

In [87]:
dp_ms['percentage_of_defaulted'].sort_values(ascending = False)

family_status
unmarried            9.763948
civil partnership    9.313014
married              7.560558
divorced             7.064760
widow / widower      6.624606
Name: percentage_of_defaulted, dtype: float64

If analyze all groups given in 'family_status' column, we got an insight that the biggest percentage of people who defaulted a loan belongs to two groups: of unmarried and civil partnership  categories. In each group approxinately 9% of people defaulted a loan.

But the meaning of these two groups('unmarried' and 'civil partnership') is completely opposite. 
So for further analysis let's check how single/divorced people return a loan in comparison with married people. For this analysis I categorized all clients into 2 groups only (column 'marital_status').

The first group includes people with married and civil partnership marital statuses.
The second group includes unmarried, divorced and  widows/widowers.

In [88]:
#this is pivot table to analyze returning a loan by two categories of people
#married/civil partnership and single

dp_ms_groupped = data.pivot_table(index=['marital_status'],columns='defaulted_a_loan', values='debt', aggfunc='count')
dp_ms_groupped['percentage_of_defaulted'] = (dp_ms_groupped['defaulted a loan before']/(dp_ms_groupped['never defaulted a loan']+dp_ms_groupped['defaulted a loan before']))*100
dp_ms_groupped['percentage_of_paying'] = 100 - dp_ms_groupped['percentage_of_defaulted']
dp_ms_groupped['ratio paying to defaulted'] = dp_ms_groupped['never defaulted a loan']/dp_ms_groupped['defaulted a loan before']

print(dp_ms_groupped)

defaulted_a_loan           defaulted a loan before  never defaulted a loan  \
marital_status                                                               
married/civil partnership                     1312                   15083   
single                                         357                    3628   

defaulted_a_loan           percentage_of_defaulted  percentage_of_paying  \
marital_status                                                             
married/civil partnership                 8.002440             91.997560   
single                                    8.958595             91.041405   

defaulted_a_loan           ratio paying to defaulted  
marital_status                                        
married/civil partnership                  11.496189  
single                                     10.162465  


In [89]:
print('Average percentage of defaulted the loan by this classification {:.2f}%'.format(dp_ms_groupped['percentage_of_defaulted'].mean()))

Average percentage of defaulted the loan by this classification 8.48%


### Conclusion

If compare only two group of people: **single and married** (where single group includes unmarried, divorced and widows/widowers, and married group includes married and civil partnership), we got an insight that there not a big difference between these two groups in returning a loan. But the **single group defaulted to pay off a loan on 0,5% often than married people.**

- An average percentage of defaulted the loan is 8.26%. 
- Where the percentage of defaulted the loan for single people is 8,5%
- And for married people is 8,0%

To provide a loan to single people according to these stats a little bit more alarming.

If looking on more detailed stats by all categories of marital status. 

- The biggest percentage of people defaulted a loan belongs to groups of **"unmarried" and "civil partnership"**.
- The percentage not return a loan among these groups approximately 9%
- **divorced and widows/widowers are the most trustable** groups of people to give them a loan.
- The percentage not to pay a loan in these groups is only 7%.

Interesting percentages of not paying are in groups:

- **"civil partnership" - 9.31%**
- **"married" - 7.56%**

The main difference between these two groups according to the google is that "marriage is formed by vows, whereas a civil partnership is formed by signing the civil partnership document". So the people who gave a vow to their partner is more trustable in paying off a loan than people who just signed their marriage at the city hall.

- **Is there a relation between income level and repaying a loan on time?**

In [90]:
#Let's build a pivot table to answer the question
#this pivot is built on 3 categories of income level
dp_il = data.pivot_table(index=['income_level'],columns='defaulted_a_loan', values='debt', aggfunc='count')

dp_il['percentage_of_defaulted'] = (dp_il['defaulted a loan before']/(dp_il['never defaulted a loan']+dp_il['defaulted a loan before']))*100
dp_il['percentage_of_paying'] = 100 - dp_il['percentage_of_defaulted']
print(dp_il)

defaulted_a_loan  defaulted a loan before  never defaulted a loan  \
income_level                                                        
greater income                        762                    8858   
median income                         170                    1922   
smaller income                        800                    8819   

defaulted_a_loan  percentage_of_defaulted  percentage_of_paying  
income_level                                                     
greater income                   7.920998             92.079002  
median income                    8.126195             91.873805  
smaller income                   8.316873             91.683127  


In [91]:
#this pivot is built on 5 categories of income level
dp_il_quartiles = data.pivot_table(index=['income_level_quartiles'],columns='defaulted_a_loan', values='debt', aggfunc='count')

dp_il_quartiles['percentage_of_defaulted'] = (dp_il_quartiles['defaulted a loan before']/(dp_il_quartiles['never defaulted a loan']+dp_il_quartiles['defaulted a loan before']))*100
dp_il_quartiles['percentage_of_paying'] = 100 - dp_il_quartiles['percentage_of_defaulted']

print(dp_il_quartiles)

defaulted_a_loan        defaulted a loan before  never defaulted a loan  \
income_level_quartiles                                                    
1st quartile                                383                    4427   
2nd quartile                                417                    4392   
3rd quartile                                423                    4387   
4th quartile                                339                    4471   
median income                               170                    1922   

defaulted_a_loan        percentage_of_defaulted  percentage_of_paying  
income_level_quartiles                                                 
1st quartile                           7.962578             92.037422  
2nd quartile                           8.671241             91.328759  
3rd quartile                           8.794179             91.205821  
4th quartile                           7.047817             92.952183  
median income                          8.1

In [92]:
#To undestand the categories better here hierarchy of them:
#smallest income - 1st quartile
#less than median - 2nd quartile
#median income
#greater than median - 3rd quartile
#greatest income - 4th quartile

dp_il_quartiles['percentage_of_defaulted'].sort_values(ascending = False)

income_level_quartiles
3rd quartile     8.794179
2nd quartile     8.671241
median income    8.126195
1st quartile     7.962578
4th quartile     7.047817
Name: percentage_of_defaulted, dtype: float64

### Conclusion

Based on the numbers above we can make an conclusion that:

- People with the biggest income (4th quartile) and people with the smallest income (1st quartile) defaulted a loan less than other people. Only 7.04% and 7.96% among them ever defaulted a loan. This is most trustable people to give tham a loan. 

- The most risky to give a loan to the people with income greater than median (3rd quartile). The percentage of not returning a loan among them is 8.79%.

This is very surprising and not obvious fact that the people with the smallest income are most trustable by returning a loan than people with median income.
This conclusion is based on classification of people to 5 groups (4 quartiles by level income and median income). 

But if take a look on the numbers when all data classified only in **three groups: greater income, median income and smaller income**, we will see a bit different result:

- That people with greatest income are the most trustable. They ever defaulted a loan in only 7,92% cases.
- And that the less trustable are the people with the smallest income. They ever defaulted a loan in 8,31% cases.

So if classify people in three categories I can say that there are a relation between income level and repaying a loan on time. The more income - the most likely the client will pay a loan on time.

But if take a look on numbers classified more detailed, we can make such an surprising conclusion like mentioned above:  that the people with the smallest income as well as people with the highest income are more trustable by returning a loan than people with median income. 

So everything depends on size and logic of classification built for the dataset.

- **How do different loan purposes affect on-time repayment of the loan?**

In [93]:
dp_purpose = data.pivot_table(index=['purpose_groupped'],columns='defaulted_a_loan', values='debt', aggfunc='count')

dp_purpose['percentage_of_defaulted'] = (dp_purpose['defaulted a loan before']/(dp_purpose['never defaulted a loan']+dp_purpose['defaulted a loan before']))*100
dp_purpose['percentage_of_paying'] = 100 - dp_purpose['percentage_of_defaulted']

print(dp_purpose)

defaulted_a_loan  defaulted a loan before  never defaulted a loan  \
purpose_groupped                                                    
car                                   400                    3879   
education                             369                    3619   
property                              780                    9971   
wedding                               183                    2130   

defaulted_a_loan  percentage_of_defaulted  percentage_of_paying  
purpose_groupped                                                 
car                              9.347978             90.652022  
education                        9.252758             90.747242  
property                         7.255139             92.744861  
wedding                          7.911803             92.088197  


In [94]:
dp_purpose['percentage_of_defaulted'].sort_values(ascending = False)

purpose_groupped
car          9.347978
education    9.252758
wedding      7.911803
property     7.255139
Name: percentage_of_defaulted, dtype: float64

### Conclusion

To answer on a question **How do different loan purposes affect on-time repayment of the loan?** I categorized the data and built a pivot table on the base of this data. So I can make an following conclusions:

1. People who took a loan to buy, renovate, or build a property (this also includes real estate and such purposes like "to buy a house for my family") has the lowest rate of defaulting a loan. This rate is 7.25%. Means that only 7.25% of the people  who took loan for real estate ever defaulted a loan.

2. The next is going people who took a loan to organize a wedding. Their default rate is 7,91%.

3. The most risky among all purposes appeared loan to buy a car with the default rate of 9.34%.

So the bank should pay more attention to the clients who want to take a loan for purchase of car. Because these people have more chances default a loan.

And maybe motivate more or advertise more loans for real estate as more defaultsafe for the bank.

### General Conclusion

What have we learned about the dataset from this analysis:

- Average default rate for the whole data set is 8.12%

- Maximum risk is to provide a loan to the client with 4 kids. Their default rate is 9,75%.
- People with 5 kids according to our data have never defaulted a loan.
- People with no kids are the most trustable for bank to provide them a loan. Their default rate is only 7.54%.
- The next group of people by trustability is people with 3 children. Their default rate is 8,18%.

- The biggest percentage of people who defaulted a loan are unmarried or in civil partnership. Their default rate is approxinately 9%.
- People who are single are most likely default a loan than married people. 
- Default rate for single people is 8,5% and for married 8,0%.
- Divorced and widows/widowers are the most trustable groups of people to give them a loan. Default rate in these groups is only 7%.
- People who got married in a church are most likely return a loan without default than people who got married in town hall or just signed the civil partnership. 

- People with the biggest income (4th income quartile) and people with the smallest income (1st quartile) defaulted a loan less than other people. 
- The most risky to give a loan to the people with income greater than median (3rd quartile). 
- This is very surprising and not obvious that the people with the smallest income are most trustable by returning a loan than people with median income. This conclusion is based on classification of people to 5 groups (4 quartiles and median income).
- But if classify the people by income level only on three groups instead of 5, the data will show that there are a relation between income level and repaying a loan on time. And that the people with greatest income are the most trustable by default rate.
- But if build categorization more detailed - we become a bit different numbers (like insight mentioned above that "people with the smallest income are most trustable by returning a loan than people with median income"). So everything depends on size of categorization groups.

- People who took a loan for operations with real estate have the lowest default rate. Only 7.25% of the people who took loan for real estate ever defaulted a loan.
- The most risky to provide a loan to the people whos purpose is to buy a car. Their default rate is 9.34%.


### Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization;
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [x]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [x]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.